pkg/kv/kvserver/replica_proposal.go

```diff
@@ -45,6 +45,70 @@ import (
 // ProposalData is data about a command which allows it to be
 // evaluated, proposed to raft, and for the result of the command to
 // be returned to the caller.
+//
+// A proposal (=ProposalData) is created and then handed to the proposal buffer
+// (Replica.mu.proposalBuf) via propBuf.Insert. In the common case, the proposal
+// transfers from there into Replica.mu.proposals (and the raft log) via
+// propBuf.FlushLockedWithRaftGroup (called from handleRaftReady), but instead
+// an additional call to Insert (on behalf of some other proposal) may also do
+// this instead (in case the propBuf fills up). ProposalData itself has no mutex
+// and locking is provided by `Replica.mu` which must always be held when
+// accessing a ProposalData (except before the first insertion into the proposal
+// buffer). However, at the time of writing, this locking is possibly insufficient
+// and the above rule is not adhered to at all times. In particular, during raft
+// application, the ProposalData is mutated without the Replica.mu lock being
+// held in all places. As log application is a well-trodden code path and no
+// relevant data races are known, they are either absent or extremely rare.
+// Still, an improvement seems advisable.
+//
+// This goes for the locking, but also for the lifecycle of ProposalData in
+// aggregate. It is extremely complex due to internal reproposals and the
+// multiple, poorly synchronized, and circular hand-offs that exist between
+// the proposals map and the proposal buffer. The following diagram attempts
+// to describe the status quo.
+//
+//	         created
+//	           │                proposal map, apply
+//	           │[1]             goroutine, and propBuf
+//	           ▼                        ▲    │
+//	┌─┬────►propBuf  ┌────────────┐     │[6] │[2]
+//	│ │        │     │            │     │    ▼
+//	│ │        │[2]  │         proposal map and
+//	│ │        ▼     ▼         apply goroutine
+//	│ │   proposal map────┐       ▲
```

Comment:
I experimented a bit and this was what I thought was most useful (for me), but I understand this may not be everyone's cup of tea. I am not terribly motivated to play comment golf, tbh, esp. since I am happy with this. Maybe "the gang" (everyone who cares) can take a stab at improving this instead? I think that would also further the understanding of those transitions across the team.

Also, I'm hoping to simplify this diagram significantly with https://github.com/cockroachdb/cockroach/pull/97779, which I've started picking up again. Not ready to commit that yet (after all, don't want to run into a time crunch in the end and there are some other tactical fixes to make), but it seems doable.

Reply:
Yeah, I don't have any strong opinions on this, but you asked for feedback. :)